src/Microsoft.Identity.Web/Resource/RequiredScopeFilter.cs

```diff
@@ -13,108 +13,108 @@
-            {
-                throw new InvalidOperationException(IDWebErrorMessage.MissingRequiredScopesForAuthorizationFilter);
-            }
+		private void ValidateEffectiveScopes(AuthorizationFilterContext context)
+		{
+			if (_effectiveAcceptedScopes == null || !_effectiveAcceptedScopes.Any())
+			{
+				throw new InvalidOperationException(IDWebErrorMessage.MissingRequiredScopesForAuthorizationFilter);
+			}
 
-            if (context.HttpContext.User == null || context.HttpContext.User.Claims == null || !context.HttpContext.User.Claims.Any())
-            {
-                context.HttpContext.Response.StatusCode = (int)HttpStatusCode.Unauthorized;
-                throw new UnauthorizedAccessException(IDWebErrorMessage.UnauthenticatedUser);
-            }
-            else
-            {
-                // Attempt with Scp claim
-                Claim? scopeClaim = context.HttpContext.User.FindFirst(ClaimConstants.Scp);
+			if (context.HttpContext.User == null || context.HttpContext.User.Claims == null || !context.HttpContext.User.Claims.Any())
+			{
+				context.HttpContext.Response.StatusCode = (int)HttpStatusCode.Unauthorized;
+				throw new UnauthorizedAccessException(IDWebErrorMessage.UnauthenticatedUser);
+			}
+			else
+			{
+				// Attempt with Scp claim
+				Claim? scopeClaim = context.HttpContext.User.FindFirst(ClaimConstants.Scp);
 
-                // Fallback to Scope claim name
-                if (scopeClaim == null)
-                {
-                    scopeClaim = context.HttpContext.User.FindFirst(ClaimConstants.Scope);
-                }
+				// Fallback to Scope claim name
+				if (scopeClaim == null)
+				{
+					scopeClaim = context.HttpContext.User.FindFirst(ClaimConstants.Scope);
+				}
 
-                if (scopeClaim == null || !scopeClaim.Value.Split(' ').Intersect(_effectiveAcceptedScopes).Any())
-                {
-                    context.HttpContext.Response.StatusCode = (int)HttpStatusCode.Forbidden;
-                    string message = string.Format(
-                        CultureInfo.InvariantCulture,
-                        IDWebErrorMessage.MissingScopes,
-                        string.Join(",", _effectiveAcceptedScopes));
-                    context.HttpContext.Response.WriteAsync(message);
-                    context.HttpContext.Response.CompleteAsync();
-                    throw new UnauthorizedAccessException(message);
-                }
-            }
-        }
+				if (scopeClaim == null || !scopeClaim.Value.Split(' ').Intersect(_effectiveAcceptedScopes).Any())
+				{
+					context.HttpContext.Response.StatusCode = (int)HttpStatusCode.Forbidden;
+					string message = string.Format(
+						CultureInfo.InvariantCulture,
+						IDWebErrorMessage.MissingScopes,
+						string.Join(",", _effectiveAcceptedScopes));
+					context.HttpContext.Response.WriteAsync(message);
+					context.HttpContext.Response.CompleteAsync();
+					throw new UnauthorizedAccessException(message);
```

Comment:
Why does an exception need to be thrown when the response was already completed in the line above?

Reply:
so the rest of the controller action is not done. if the scope is not correct, we want to avoid the rest of the controller action from happening. there might be a better way to handle it, will take a look. if you have any suggestions, let us know.